### PR TITLE
Type inference profiling

### PR DESCRIPTION
**Pull request by vilterp**
TODO:
- [x] make it work with new API that returns an array instead of a single tree
- [x] figure out why the profiles I'm getting in test are empty
- [x] add tests

### CODE REVIEW

File: src/ProfileEndpoints.jl
Comment (vilterp) [inline src/ProfileEndpoints.jl]: I'd like to just spoof a root `InferenceTimingNode` here instead of changing `flamegraph` downstream to work with the array of trees, if possible

**Comment by vilterp** [comment]
Error in CI:
```
│    InexactError: trunc(Int64, 1.8446663302938247e19)
│    Stacktrace:
│      [1] trunc
│        @ ./float.jl:781 [inlined]
│      [2] round
│        @ ./float.jl:359 [inlined]
│      [3] _flamegraph_frame(io::IOBuffer, node::SnoopCompileCore.InferenceTimingNode, start_secs::Float64, check_precompilable::Bool, excluded_modules::Set{Module}, mode::Nothing; toplevel::Bool)
│        @ SnoopCompile ~/.julia/packages/SnoopCompile/Q8zUg/src/parcel_snoopi_deep.jl:1863
│      [4] flamegraph(tinf::SnoopCompileCore.InferenceTimingNode; tmin::Float64, excluded_modules::Set{Module}, mode::Nothing)
│        @ SnoopCompile ~/.julia/packages/SnoopCompile/Q8zUg/src/parcel_snoopi_deep.jl:1792
│      [5] flamegraph
│        @ ~/.julia/packages/SnoopCompile/Q8zUg/src/parcel_snoopi_deep.jl:1788 [inlined]
│      [6] typeinf_stop_endpoint(req::HTTP.Messages.Request)
│        @ ProfileEndpoints ~/work/ProfileEndpoints.jl/ProfileEndpoints.jl/src/ProfileEndpoints.jl:265
```

In some Julia versions but not others? Wut? 🤔

File: src/ProfileEndpoints.jl
Comment (NHDaly) [reply src/ProfileEndpoints.jl]: Yeah, i think that's the right approach. i ended up doing that too in the snoopcompile PR, here:
https://github.com/timholy/SnoopCompile.jl/pull/313#discussion_r1059492749

File: src/ProfileEndpoints.jl
Comment (NHDaly) [inline src/ProfileEndpoints.jl]: For the old version, where `clear_and_fetch_timings` isn't defined, i think you also need to call reset here.

What i would probably recommend is that you directly call `SnoopCompileCore.start_deep_timing()` instead, since in https://github.com/timholy/SnoopCompile.jl/pull/313, i've changed those functions to always be defined correctly whether it's the new API or the old one.

File: src/ProfileEndpoints.jl
Comment (NHDaly) [inline src/ProfileEndpoints.jl]: Same here actually -- i think you can directly call `SnoopCompileCore. stop_deep_timing(); SnoopCompileCore.finish_snoopi_deep()`, instead of having to rebuild it yourself, and this way the different API logic is contained to just that one package and you don't need to check `isdefined` stuff.

File: Project.toml
Comment (NHDaly) [inline Project.toml:12]: 🤔 I remember that we wanted to avoid having SnoopCompile in the dependencies for our server, because it adds a bunch of expensive deps for Cthulhu and FlameGraphs and stuff...

I'm thinking maybe what we could do is just directly return the InferenceTimingNodes, and then have the _client_ do the conversion to PProf? You could send the profiles directly in their original form, via `Serialization.serialize()`, and then reconstruct them into pprof profiles on the client side? I think it would be okay to add these deps to our benchmarking package, rather than to our server.

File: src/ProfileEndpoints.jl
Comment (NHDaly) [inline src/ProfileEndpoints.jl]: i don't know why we did all of these this way, but it seems a bit silly now...

Can we maybe switch them all to `@assert precompile(...)` instead of `precompile(...) || error(...)`? It's shorter, easier to read, and easier to copy/paste. If you don't want to do that change in this PR tho, we can do it later 👍

File: test/runtests.jl
Comment (NHDaly) [inline test/runtests.jl]: if we changed this to return the original profile data, not the pprof flamegraph, it might be easier to test as well, since `foo` should be one of the top-level inference invocations.

**Review by NHDaly — COMMENTED** [review]
@vilterp: This is _such a good idea_! I'm really glad you thought of this - we should have started with this from the beginning! 💡 

It'll make it so much easier to grab profiles, and it'll let us farm out the work to anyone interested, and it can unblock the work we did so that at least we have _something_ enabled. Great idea. :)

I'd love to also do it for LLVM opt profiling - maybe we can do that in a followup once this one is merged. Thanks again, great idea 😊

**Comment by NHDaly** [comment]
Huh yeah i'm not sure how to explain the CI failure.. :/ it should go away if we move the flamegraph conversion out of this package and then we can maybe restrict our client code to a working version of the SnoopCompile package if that's the problem

File: Project.toml
Comment (NHDaly) [reply Project.toml:12]: 🤔 but actually, looking at this dependencies list, this doesn't seem sooo bad:
```
(ProfileEndpoints) pkg> add SnoopCompile FlameGraphs
   Resolving package versions...
    Updating `~/.julia/dev/ProfileEndpoints/Project.toml`
⌅ [08572546] + FlameGraphs v0.2.10
  [aa65fe97] + SnoopCompile v2.9.6
    Updating `~/.julia/dev/ProfileEndpoints/Manifest.toml`
  [da1fd8a2] + CodeTracking v1.2.0
  [f68482b8] + Cthulhu v2.7.6
  [1eca21be] + FoldingTrees v1.2.1
  [c3a54625] + JET v0.6.21
  [aa1ae85d] + JuliaInterpreter v0.9.20
  [6f1432cf] + LoweredCodeUtils v2.3.0
  [1914dd2f] + MacroTools v0.5.10
  [295af30f] + Revise v3.5.0
  [aa65fe97] + SnoopCompile v2.9.6
  [69024149] + StringEncodings v0.3.5
  [ddb6d928] + YAML v0.4.8
```

🤔 Do you remember why we wanted to avoid SnoopCompile as a dependency? 🤔

Comment (NHDaly) [reply Project.toml:12]: Aha! I found the culprit -- it turned out it was actually an unrelated change in your PR, coming from PProf!
https://github.com/RelationalAI/raicode/pull/10713/files#diff-6d1205e03b7a43c01ed2ac8b8c83eb58c6f3a3b1360769e18644245abee8c565

It's this issue, here:
https://github.com/JuliaPerf/PProf.jl/issues/72, which we're now tracking here as well: https://relationalai.atlassian.net/browse/RAI-8826.

So i think this is actually maybe fine? 👍 
It'd still be nice to not have to depend on Revise, JET, and Cthulhu at runtime, though, if we could help it. So it might still be worth pulling this out and doing this work client-side.

Comment (NHDaly) [reply Project.toml:12]: Okay drats, i tried this locally, but it actually doesn't work, since the inference profile directly references all the methods that were profiled, and those methods won't exist on the client. For example:
```julia
julia> deserialize("/Users/nathandaly/Downloads/inference_profile.jlserialized")
ERROR: KeyError: key HTTP [cd3eb016-35fb-5094-929b-558a96fad6f3] not found
```
So either we'd need to convert all those method references to a string, and build a new profile object (flamegraph), or we just do what you have here and import SnoopCompile to do that for us. 👍 

So I guess I'd say we leave this as-is, for now. 👍

Comment (NHDaly) [reply Project.toml:12]: PR for the above, here: https://github.com/JuliaPerf/ProfileEndpoints.jl/pull/17, which i immediately closed.

**Comment by NHDaly** [comment]
Ah okay actually, i think the test failure was because you weren't finishing the cleanup on the ROOT timing node in your version. It's fixed now, because we switched to just calling the functions directly from SnoopCompile. I think this is good to go then!

**Review by NHDaly — APPROVED** [review]
Thanks @vilterp! LGTM

**Pull request merged**
